### DISCUSSION TITLE
Set default code in frontend for lock entity

### DISCUSSION THREAD

**Discussion by gjohansson-ST** [home-assistant/architecture#870] [Entity Models] [answered]
from [PR](https://github.com/home-assistant/core/pull/87544#discussion_r1097981448) it is currently in current implementation not possible to use a default code.

### Problem

Implementation of validation function in `lock` has removed functionality to use a default `code` provided by the user in the integration and users now need to always provide a `code` if `code_format` is set.

### Proposal

- Implement a user customization for `lock` so that a user can set a default code for a certain `lock` entity in frontend.
- Use the default `code` as fallback in the validation from the entity registry for the specific `lock` entity.
- If no `code` is provided by the user and no fallback default `code` is set then fail validation if `code_format` is set (current implementation).

### Issues with suggestion

- Integrations using a default code supplied from config entry data should be rewritten to remove this option

### Result

Users can now directly in frontend for the specific `lock` entity provide a default `code` which will be considered in validation and thereby `code` can be omited in service calls even if `code_format` has been set to validate the `code`.

### Reference PR's
Frontend PR: https://github.com/home-assistant/frontend/pull/15450
Backend PR: https://github.com/home-assistant/core/pull/88139

**Comment by jbouwh** [comment]
I would not suggest storing the code in the entity. Either use the service call with an automation, or have a dialog to enter the code. 
Storing the code is something to do at integration level. Adding a property `default_code` does not make sense.

**Reply by jbouwh** [reply]
I would suggest adding an option to make code validation default. (https://github.com/home-assistant/architecture/discussions/872)

**Comment by balloob** [comment]
How would an integration know what default code to provide? It seems like a couple of things are conflated here and this would be an entity option for the lock entity instead. Even then you end up maybe having to provide codes for each different action?

**Reply by gjohansson-ST** [reply]
Most integrations that I have seen so far (`yale_smart_alarm`, `verisure` as examples) allows the user to store a default code as config entry options while obviously it could be provided by YAML for integrations etc. (`MQTT` could be one example).

So the idea is quite simple to allow the integration to set (as example) `_attr_default_code = config_entry.options[CONF_CODE]` to also allow a default code to be tested in the validation and the user would not have to provide a code in frontend.

In my way this should solve all the use-cases so if the integration expands to allow saving a default code from the user it can be used in validation. If not code is required for the integration then the developer should not set `_attr_code_format` at all and the validation will not be used.

For integrations that requires a `code` for `unlock` but not for `lock` (example: `yale_smart_alarm`) I believe we can leave it as on/off and then `code` could be required for `lock` as well. If `_attr_default_code` is implemented it would be simple enough for the user experience.

I hope this makes sense.

**Reply by balloob** [reply]
Device Integrations should only integrate APIs and nothing more. This is clearly a feature on top and so should not be provided by device integrations. If offered, it's the Lock integration itself that should offer this (the Lock entity).

**Reply by gjohansson-ST** [reply]
Yes. Maybe I'm not using the right terminology but I'm suggesting to expand the lock entity component including it's validation to allow developers to offer the user to set a default code.

For clarify I want to expand [LockEntity](https://github.com/home-assistant/core/blob/dev/homeassistant/components/lock/__init__.py#L135) with a new property which is `default_code`

**Reply by balloob** [reply]
Right, but that default code should never be set by the integration itself. As it’s user provided data. The Yale integration only should provide data from the Yale API, not things that the user can enter. If default lock needs to be a thing, it’s a feature that needs to be provided by the `lock` integration for a `lock` entity, regardless of the platform that provides that entity. No code should be added for default code to any integration that provides lock entities because it’s a generic user configuration.

**Reply by gjohansson-ST** [reply]
For clarity I made a PR what is my suggestion https://github.com/home-assistant/core/pull/88012
Would this be ok from an implementation perspective?

**Reply by balloob** [reply]
No. As this requires the integrations that integrate an API to provide the default lock code which is explicitly saying we should not do. It's not their responsibilitty.

**Reply by gjohansson-ST** [reply]
I'm not sure I understand as this is the concept which is applied to many integrations today and integrations does come with a config flow to get user input.

Anyhow if that is a no-go could we consider instead a concept of storing a default code from the user by expanding the frontend (settings dialog) for the `lock` entity and store a default `code` in the entity registry to be used?
Perhaps that is an even better proposal and more in line what is happening on others.
I would then adjust my proposal accordingly.

**Reply by balloob** [reply]
What you're proposing now is what I have been saying from the start … if the lock integration handles it, that's indeed where it would be stored. Just like sensor integration stores sensor precision that way.

**Reply by gjohansson-ST** [reply]
I have updated my proposal accordingly. Let me know if you see any particular adjustments but otherwise I meant more or less exactly as how `precision` is handled for `sensor`.

**Reply by gjohansson-ST** [reply]
Happy with the new proposal @balloob ? If so should probably open another discussion for adding this also for `alarm_control_panel`

**Comment by emontnemery** [comment, marked as answer]
The proposal is good, let's go ahead with core + frontend implementation 👍
